src/video/SDL_surface.c

```diff
@@ -139,19 +154,36 @@ SDL_Surface *SDL_CreateSurface(int width, int height, SDL_PixelFormatEnum format
         return NULL;
     }
 
-    surface->format = SDL_CreatePixelFormat(format);
-    if (!surface->format) {
-        SDL_DestroySurface(surface);
-        return NULL;
-    }
+    surface->format = format;
     surface->w = width;
     surface->h = height;
     surface->pitch = (int)pitch;
-    SDL_SetSurfaceClipRect(surface, NULL);
 
-    if (SDL_ISPIXELFORMAT_INDEXED(surface->format->format)) {
-        SDL_Palette *palette =
-            SDL_CreatePalette((1 << surface->format->bits_per_pixel));
+    surface->internal = (SDL_SurfaceData *)SDL_calloc(1, sizeof(*surface->internal));
+    if (!surface->internal) {
+        SDL_DestroySurface(surface);
+        return NULL;
+    }
+
+    surface->internal->format = SDL_GetPixelFormatDetails(format);
+    if (!surface->internal->format) {
+        SDL_DestroySurface(surface);
+        return NULL;
+    }
+
+    /* Initialize the clip rect */
+    surface->internal->clip_rect.w = width;
+    surface->internal->clip_rect.h = height;
+
+    /* Allocate an empty mapping */
+    surface->internal->map = SDL_AllocBlitMap();
```

Comment:
Can the blitmap be stored by value in the surfacedata?

Reply:
Yes, but it requires some tap dancing with structure definitions. I can do that.